apps/survey/contracts/Survey.sol

```diff
@@ -0,0 +1,250 @@
+        MultiOptionVote storage previouslyVoted = survey.voters[msg.sender];
+
+        // revert previous votes, if any
+        if (previouslyVoted.optionsCastedLength > 0) {
+            // voter removes their vote
+            for (uint256 i = 1; i <= previouslyVoted.optionsCastedLength; i++) {
+                survey.votes[previouslyVoted.castedVotes[i].optionId] = survey.votes[previouslyVoted.castedVotes[i].optionId].sub(previouslyVoted.castedVotes[i].stake);
+            }
+            // remove previous vote participation
+            survey.participation = survey.participation.sub(voterStake.sub(previouslyVoted.castedVotes[0].stake));
+
+            // reset previously voted options, if any
+            delete survey.voters[msg.sender];
+        }
+
+        uint256 totalVoted = 0;
+        // reserve first index for NO_VOTE
+        survey.voters[msg.sender].castedVotes[0] = OptionCast({optionId: NO_VOTE, stake: 0});
+        for (uint256 j = 0; j < _options.length; j++) {
+            require(_options[j] != NO_VOTE && _options[j] <= survey.options);
+            require(_stakes[j] > 0);
+
+            // register voter amount
+            // index 0 in options array is reserved for NO_VOTE, so index will be 1 more
+            survey.voters[msg.sender].castedVotes[j + 1] = OptionCast({optionId: _options[j], stake: _stakes[j]});
+
+            // add to total option support
+            survey.votes[_options[j]] = survey.votes[_options[j]].add(_stakes[j]);
+
+            // let's avoid repeating an option by
+            // making sure that ascending order is preserved in options array
+            // index 0 in options array is reserved for NO_VOTE, so here we are comparing
+            // the option just added (_options[j] = options[j+1]) with the previous one (options[j])
+            require(survey.voters[msg.sender].castedVotes[j].optionId < survey.voters[msg.sender].castedVotes[j + 1].optionId);
+            totalVoted = totalVoted.add(_stakes[j]);
+
+            CastVote(_surveyId, msg.sender, _options[j], _stakes[j]);
+        }
+
+        // compute and register non used tokens
+        // implictly we are doing require(totalVoted <= voterStake) too
+        // (as stated before, index 0 is for NO_VOTE option)
+        survey.voters[msg.sender].castedVotes[0].stake = voterStake.sub(totalVoted);
+
+        // register number of options voted
+        survey.voters[msg.sender].optionsCastedLength = _options.length;
+
+        // add voter tokens to participation
+        survey.participation = survey.participation.add(totalVoted);
+    }
+
+    /**
+    * @notice Vote option #`_option` in survey #`_surveyId`.
+    * @dev It will use the whole balance. If voting option is 0, that's like cancelling the previously casted vote
+    * @param _surveyId Id for survey
+    * @param _option Index of supported option
+    */
+    function voteOption(uint256 _surveyId, uint256 _option) public {
```

Comment:
Any time where we use `_option`, I think it'd be more clear if we used `_optionId`.